### PR TITLE
 Add an option to overwrite the existing checkpoint file

### LINKED ISSUE #15106: ModelCheckpoint callback should [have the option to] override existing .ckpt when name matches existing file
## 🚀 Feature

Option for ModelCheckpoint callback to override a .ckpt file if one of the same filename already exists, instead of modifying the name and storing both ckpt. This could be something like a new param in the ModelCheckpoint callback override_existing=False.

### Motivation
Currently if i provide a filename that already exists, the callback automatically appends a -v1 to the name. For cases when we rely on naming conventions for ckpt files, and we try to enforce those names by passing a specific name to `ModelCheckpoint.filename`, we'd like the .ckpt to have that exact name, not some modified version of it

### Pitch

I propose a new parameter in the ModelCheckpoint callback that by default could match current behavior, but give the option to override any existing ckpt file of the same name if it already exists. This could be `override_existing=False` on the ModelCheckpoint constructor.

### Alternatives

- current solution which appends a suffix to name breaks pipelines that do stuff with the ckpt afterwards and rely on the name matching what was provided to Modelcheckpoint.filename parameter.
- Making my proposal the default behavior instead of an option: considered this but would be less impactful to keep it as an option that, if left untriggered, maintains current behavior.

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

### PR DESCRIPTION
**Pull request by solyaH**
## What does this PR do?

If you want to preserve the checkpoint's file name, you can now specify `override_existing=True` on the ModelCheckpoint constructor. It will lead to overwriting the file path instead of writing a new file with the version appended.
Fixes #15106

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/pytorch/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -129,6 +129,8 @@ class ModelCheckpoint(Checkpoint):
             where both values for ``every_n_epochs`` and ``check_val_every_n_epoch`` evenly divide E.
         save_on_train_epoch_end: Whether to run checkpointing at the end of the training epoch.
             If this is ``False``, then the check runs at the end of the validation.
+        override_existing: Whether to overwrite the checkpoint file.
```
Comment (awaelchli) [inline src/lightning/pytorch/callbacks/model_checkpoint.py:132]: Did you mean overwrite? Also, the expression "overwrite existing" is probably redundant, since overwriting can only be done if something exists, otherwise it would just be called "writing".

File: tests/tests_pytorch/checkpointing/test_model_checkpoint.py [hunk 2]
```diff
@@ -1171,6 +1171,59 @@ def test_ckpt_version_after_rerun_same_trainer(tmpdir):
     assert set(os.listdir(tmpdir)) == expected
 
 
+def test_ckpt_overriding_enabled_after_rerun_new_trainer(tmpdir):
+    """Check that previous checkpoints are renamed to have the correct version suffix when new trainer instances
+    are used."""
```
Comment (awaelchli) [inline tests/tests_pytorch/checkpointing/test_model_checkpoint.py:1176]: I think that with the feature enabled, there aren't any suffixes? And there is no renaming. 
```suggestion
    """Check that previous checkpoints get overwritten an no suffixes are generated."""
```

File: src/lightning/pytorch/callbacks/model_checkpoint.py (hunk 1 above)
Comment (solyaH) [reply src/lightning/pytorch/callbacks/model_checkpoint.py:132]: I've followed the suggested option in the discussion. Do you find `overwrite_existing` more suitable?
The purpose of this flag is exactly to reuse passed filepath if such exists. This is mainly useful when you have some pipeline run on top of the training and you need your checkpoint to end up in some specific location.

File: tests/tests_pytorch/checkpointing/test_model_checkpoint.py (hunk 2 above)
Comment (solyaH) [reply tests/tests_pytorch/checkpointing/test_model_checkpoint.py:1176]: Missed to change description, my bad.

File: src/lightning/pytorch/callbacks/model_checkpoint.py (hunk 1 above)
Comment (Borda) [reply src/lightning/pytorch/callbacks/model_checkpoint.py:132]:
```suggestion
        overwrite_existing: Whether to overwrite the checkpoint file.
```

**Review by Borda — APPROVED** [review]

File: src/lightning/pytorch/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -129,6 +129,8 @@ class ModelCheckpoint(Checkpoint):
             where both values for ``every_n_epochs`` and ``check_val_every_n_epoch`` evenly divide E.
         save_on_train_epoch_end: Whether to run checkpointing at the end of the training epoch.
             If this is ``False``, then the check runs at the end of the validation.
+        overwrite_existing: Whether to overwrite the checkpoint file.
+            If this is ``True``, then the name of the saved file won't be appended with a version.
```
Comment (carmocca) [inline src/lightning/pytorch/callbacks/model_checkpoint.py:133]: I propose that we are explicit here and name this `enable_version_counter` (or similar)

overwrite could refer to different things (the input checkpoint, the output checkpoints, top_k checkpoints, save_last checkpoints, ...)

Comment (solyaH) [reply src/lightning/pytorch/callbacks/model_checkpoint.py:133]: @carmocca What else bothers me: besides the best checkpoint, the versioning approach is also applied to the [last one](https://github.com/Lightning-AI/lightning/blob/6b7bc421f26a724cf97c2b9d0109d583a0592972/src/lightning/pytorch/callbacks/model_checkpoint.py#L648). So if we stick to `enable_version_counter` does it make sense to update the `_save_last_checkpoint` as well?

Comment (solyaH) [reply src/lightning/pytorch/callbacks/model_checkpoint.py:133]: I guess that there is no need to separate them (at least for now, I can't come up with a case when we would need such functionality). You either overwrite files in the directory or create new ones by versioning.

Comment (carmocca) [reply src/lightning/pytorch/callbacks/model_checkpoint.py:133]: It's a stretch but maybe you want to version "top-k" checkpoints but not "last" checkpoints

Comment (solyaH) [reply src/lightning/pytorch/callbacks/model_checkpoint.py:133]: Ok, do you find `enable_top_k_versioning`, `enable_last_versioning` reasonable names for such params?

Comment (carmocca) [reply src/lightning/pytorch/callbacks/model_checkpoint.py:133]: No, just have one param that applies to both. For those who want to apply it to one, they'd do:

```python
mc1 = ModelCheckpoint(save_top_k=5, enable_version_counter=True)
mc2 = ModelCheckpoint(save_last=True, enable_version_counter=False)
trainer = Trainer(callbacks=[mc1, mc2])
```

Comment (solyaH) [reply src/lightning/pytorch/callbacks/model_checkpoint.py:133]: Oh, I see, that makes sense. I'll go with implementing `enable_version_counter`. Thanks!

File: tests/tests_pytorch/checkpointing/test_model_checkpoint.py [hunk 4]
```diff
@@ -1320,6 +1375,23 @@ def test_save_last_versioning(tmpdir):
     assert {"last.ckpt", "last-v1.ckpt"} == set(os.listdir(tmpdir))
 
 
+def test_save_last_versioning_disabled(tmpdir):
```
Comment (carmocca) [inline tests/tests_pytorch/checkpointing/test_model_checkpoint.py:1378]: This test could be removed in favor of also adding `save_last=True` to `test_ckpt_version_counter_disabled_after_rerun_new_trainer`

**Review by carmocca — COMMENTED** [review]
Can you add a CHANGELOG.md entry, please?

File: tests/tests_pytorch/checkpointing/test_model_checkpoint.py [hunk 5]
```diff
@@ -1171,6 +1171,61 @@ def test_ckpt_version_after_rerun_same_trainer(tmpdir):
...
+        # check best_k_models state
+        expected = {"epoch=0.ckpt", "epoch=1.ckpt"}
+        assert {Path(f).name for f in mc.best_k_models} == expected
+
+    # check created ckpts
+    actual = {f.basename for f in tmpdir.listdir()}
+    assert actual == {"epoch=0.ckpt", "epoch=1.ckpt"}
+
+
+def test_ckpt_version_counter_disabled_after_rerun_same_trainer(tmpdir):
```
Comment (carmocca) [inline tests/tests_pytorch/checkpointing/test_model_checkpoint.py:1203]: What value does this test add that isn't added by the one above?

Comment (solyaH) [reply tests/tests_pytorch/checkpointing/test_model_checkpoint.py:1203]: I thought that it might be useful to check with the same trainer. But after all, it looks redundant. Removing.

**Review by carmocca — APPROVED** [review]
Awesome!
